components/tor/tor_launcher_factory.cc

```diff
@@ -313,8 +334,12 @@ void TorLauncherFactory::GotCircuitEstablished(bool error, bool established) {
     return;
   }
   is_connected_ = established;
-  for (auto& observer : observers_)
-    observer.OnTorCircuitEstablished(established);
+  if (!is_connected_) {
+    return;
+  }
+  for (auto& observer : observers_) {
+    observer.OnTorCircuitEstablished(true);
```

Comment:
Isn't this needed for `BravePrivateNewTabPageHandler::OnTorCircuitEstablished` for showing `ConnectionStatus::kConnectionStuck`? If that situation can't be reached anymore and we only want observer to know successful circuit established event, then I think we can change `OnTorCircuitEstablished(bool)` to just `OnTorCircuitEstablished()`

Reply:
GotCircuitEstablished only used to get the current connection state, so if it is connected then launcher factory won't notify us about the connection progress (because result is already known), otherwise the connection process starts and will be finished with OnTorCircuitEstablished(true or false). If OnTorCircuitEstablished(false) is not cutted off from GotCircuitEstablished then consumers started from some kind of _error_ state, but this is wrong because launcher didn't try to connect yet.

Reply:
I see, I just recalled I wrote this
```
  // A Circuit might have been established when Tor control is ready, in that
  // case we will not receive circuit established events. So we query the status
  // directly as fail safe, otherwise Tor window might stuck in disconnected
  // state while Tor circuit is ready.
  control_->GetCircuitEstablished(base::BindPostTask(
      base::SequencedTaskRunner::GetCurrentDefault(),
      base::BindOnce(&TorLauncherFactory::GotCircuitEstablished,
                     weak_ptr_factory_.GetWeakPtr())));
```
And we rely on `CIRCUIT_ESTABLISHED` and `CIRCUIT_NOT_ESTABLISHED` events to reflect the final circuit state.